client/client.go

```diff
@@ -225,17 +235,22 @@ func (c *Client) Activate(ctx context.Context) error {
 		return err
 	}
 
+	c.clientMutex.Lock()
 	c.status = activated
```

Comment:
Yorkie's client status is also managed by the DB.
I am currently concerned about the range of Mutex. This is because I expect that the client state of the actual DB and the state of the Client struct may be different.

The situation I assumed is the following:
![image](https://user-images.githubusercontent.com/40932237/133879083-5f31a09c-86f7-45bd-9c0b-052630ebc034.png)


So, I think that even if you call Activate and Deactivate in multiple routines, you need a test in which the database and instance state match.
What do you guys think?

Reply:
Yes, I think I need some Activate and Deactivate test in goroutines, and check the final condition. But the problem is, that even if I wait for all goroutines in activate/deactivate in that test, I'm not sure that it guarantees the client's state (ActivateClient or so on). Waiting for the goroutines is not enough. This makes me really hard even the other test cases that I try to append (like attach/detach test and Watch test). gRpc side is a whole black box in this test. I know that this gRpc itself is concurrent, but doubt it make this case to be solved.

Reply:
I sympathize with the problem you are experiencing. Let's discuss it together and solve it one by one. :)
The integration test is a black-box test, so there is no need to access the low-level (db) and compare directly.
Instead, it seems to be possible to compare using the API used by the user.
I wrote a temporary test to reproduce the problem I mentioned.

For Example:
```go
        MAX_ROUTINES := 5

	t.Run("concurrent activate/deactivate test", func(t *testing.T) {
		cli, err := client.Dial(defaultAgent.RPCAddr())
		assert.NoError(t, err)
		defer func() {
			err := cli.Close()
			assert.NoError(t, err)
		}()

		ctx := context.Background()
		wg := gosync.WaitGroup{}
		doc := document.New(helper.Collection, t.Name())

		// Activate <-> Deactivate in goroutine
		for i := 0; i <= 50; i++ {
			for i := 1; i <= MAX_ROUTINES; i++ {
				wg.Add(1)
				go func(idx int) {
					defer wg.Done()
					time.Sleep(time.Duration(rand.Intn(5)+4) * time.Millisecond)
					if idx%2 == 0 {
						assert.NoError(t, cli.Deactivate(ctx))
						return
					}
					assert.NoError(t, cli.Activate(ctx))
				}(i)
			}

			wg.Wait()
			if cli.IsActive() {
				assert.NoError(t, cli.Attach(ctx, doc))
				assert.NoError(t, cli.Detach(ctx, doc))
				assert.NoError(t, cli.Deactivate(ctx))
			}
		}
	})
```

In this test, using the `cli.Attach()` API to test the final condition of a client.
And this test currently throws the following failure:

```bash
.........
2021-09-25T17:24:32.550+0900	WARN	RPC : "/api.Yorkie/AttachDocument" 3.684915ms: "client_id:\"aN\\334\\300\\027\\220\\261\\232S\\347[V\" change_pack:<document_key:<collection:\"test-collection\" document:\"TestClientConcurrency/concurrent_activate/deactivate_test\" > checkpoint:<> > " => "client not activated"
2021-09-25T17:24:32.550+0900	ERROR	rpc error: code = FailedPrecondition desc = client not activated
github.com/yorkie-team/yorkie/client.(*Client).Attach
	/Users/scott/Workspace/dc7303/yorkie/client/client.go:303
command-line-arguments.TestClientConcurrency.func1
	/Users/scott/Workspace/dc7303/yorkie/test/integration/client_concurrency_test.go:67
testing.tRunner
	/Users/scott/.gvm/gos/go1.17/src/testing/testing.go:1259
2021-09-25T17:24:32.552+0900	INFO	RPC : "/api.Yorkie/DeactivateClient" 1.181214ms
--- FAIL: TestClientConcurrency (0.02s)
    --- FAIL: TestClientConcurrency/concurrent_activate/deactivate_test (0.02s)
        client_concurrency_test.go:67:
            	Error Trace:	client_concurrency_test.go:67
            	Error:      	Received unexpected error:
            	            	rpc error: code = FailedPrecondition desc = client not activated
            	Test:       	TestClientConcurrency/concurrent_activate/deactivate_test
FAIL
2021-09-25T17:24:32.556+0900	INFO	backend stoped: id: c57dpg6a0brql6r6i91g, rpc: localhost:21201
FAIL	command-line-arguments	0.313s
FAIL
```

It's not always a failing test. Because testing concurrency can sometimes be inconsistent.
Still, through these tests, I found that the issue I mentioned earlier can be reproduced and that this is a situation that users may encounter while using it.
The issue of adding this inconsistent test remains to be discussed further, but it is meaningful to know and deal with this situation.

How about improving by testing the API one after the other while writing the integration test from the user's point of view? After resolving `activate/deactivate`, I think that it will be possible to complete it while thinking about `attach/detach`, `watch`, etc. separately.

Reply:
Thanks! I really look forward to discuss this on sync time.